lazy (#179)
r-hub/blog: content/post/2025-02-01-lazy/index.Rmd

          sep = " "
        )
      },
      ">}}\n"
    )
  }
)

```

```{r, include = FALSE, cache = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  out.width = "100%",
  asciicast_theme = "pkgdown"
)
```

In the programming world, laziness can often be a good thing: it is both a human quality that can motivate automation efforts, and a programming concept that avoids wasting resources such as memory.
Now, when reading code or documentation, seeing the word "lazy" can be confusing, because of its polisemy: it carries several meanings.
In this post, we will enumerate the different possible definitions of "lazy" in R code.

## Lazy as in lazy evaluation

You might know that R provides lazy evaluation: the arguments of a function are only evaluated if they are accessed.
In short, you can pass anything as an argument value to a function without any problem as long as the function does not use that value.
The contrary of lazy evaluation is *eager evaluation*.

The [Advanced R book by Hadley Wickham](https://adv-r.hadley.nz/functions.html#lazy-evaluation) features a very clear introduction to lazy evaluation.

> Note that the workhorse of lazy evaluation in base R is a thing called a promise that contains an expression (the recipe for getting a value), an environment (the ingredients that are around), a value.

Reviewer: maybe single quote or emphasise 'promise' to make the word stand out as a term?

The latter is only computed when accessed, and cached once computed.

### What about {future}?

Maybe you have heard of the [future package](https://future.futureverse.org/index.html) by Henrik Bengtsson.
It provides an implementation in R of futures, a programming concept.
Its homepage state "In programming, a future is an abstraction for a value that may be available at some point in the future. The state of a future can either be unresolved or resolved."

> With futures, you create a future, that is associated to a promise, not the same as the aforementioned base R promise, which is a placeholder for a value and then the value itself.

Reviewer: this is a very heavy sentence, mainly because of the concepts. I think maybe placing the clarifying part-sentence in paretheses might make it better, as it would make it more clear that the last part of the sentence is an explanation of what a 'promise' is in this context (rather than it now possibly being interpreted as what the aforementioned 'promise' was).

Author: adding an example and rephrasing

The value can be computed asynchronously, which means in parallel.
Therefore, the futures package allows R programmers to take full advantage of their local computing resources: cores, clusters, etc.

Now, to add confusion beyond the different meaning of "promise" in this context, by default a future is **not lazy**, it is **eager**.
This means that it is computed immediately unless you specify an alternative method (a "plan").

> Therefore, the use of futures and the use of lazy evaluation are orthogonal concepts.

Reviewer: orthogonal is a great description, but I think also quite complicated as a word, maybe elaborate what it means in this context

  asciicast_theme = "pkgdown"
)
```

In the programming world, laziness can often be a good thing: it is both a human quality that can motivate automation efforts, and a programming concept that avoids wasting resources such as memory.
Now, when reading code or documentation, seeing the word "lazy" can be confusing, because of its polisemy: it carries several meanings.
In this post, we will enumerate the different possible definitions of "lazy" in R code.

## Lazy as in lazy evaluation

You might know that R provides **lazy evaluation**: the arguments of a function are only evaluated if they are accessed.
In short, you can pass anything as an argument value to a function without any problem as long as the function does not use that value.

For instance, the code below works despite `evaluation`'s not existing, because the definition of the `mean()` function includes ellipsis, and because the `lazy` argument is actually not used.

```{r}
mean(1:10, lazy = evaluation)
```

The contrary of lazy evaluation is **eager evaluation**.

The [Advanced R book by Hadley Wickham](https://adv-r.hadley.nz/functions.html#lazy-evaluation) features a very clear introduction to lazy evaluation.

> Note that the workhorse of lazy evaluation in base R is a thing called a **promise** that contains an *expression* (the recipe for getting a value), an *environment* (the ingredients that are around), a *value*.

Reviewer 2: ```suggestion
Note that the workhorse of lazy evaluation in base R is a thing called a **promise** that contains an *expression* (the recipe for getting a value), an *environment* (the ingredients that are around), and a *value*.
```
That missing 'and' really threw for me an unexpected loop on the first read 😅

The value can be computed asynchronously, which means in parallel.
Therefore, the futures package allows R programmers to take full advantage of their local computing resources: cores, clusters, etc.

Now, to add confusion beyond the different meaning of "promise" in this context, [by default](https://future.futureverse.org/reference/future.html) a future is **not lazy**, it is **eager**.
This means that it is computed immediately.

By default, the creation of a future below (`eager_future`) takes as much time as not wrapping the code in a future, because the computation is immediate.
Setting `lazy` to `TRUE` makes the future creation much faster (`lazy_future`).

```{r, cache=TRUE}
library("future")
bench::mark(no_future <- is.numeric(runif(n = 10000000)))
bench::mark(eager_future <- future(is.numeric(runif(n = 10000000))))
bench::mark(lazy_future <- future(is.numeric(runif(n = 10000000)), lazy = TRUE))
```

If we do retrieve the value, overall the same time is spent between creating the future and our getting the value:

```{r, cache=TRUE, warning=FALSE}

> bench::mark({
>   no_future <- is.numeric(runif(n = 10000000))
>   no_future
> })
> 
> bench::mark({
>   eager_future <- future(is.numeric(runif(n = 10000000)))
>   value(eager_future)
> })
> 
> bench::mark({
>   lazy_future <- future(is.numeric(runif(n = 10000000)), lazy = TRUE)
>   value(lazy_future)
> })

Reviewer 2: ```suggestion
bench::mark(
  no_future = {is.numeric(runif(n = 10000000))},
  eager_future = {x <- future(is.numeric(runif(n = 10000000))); value(x)},
  lazy_future = {x <- future(is.numeric(runif(n = 10000000)), lazy = TRUE); value(x)},
  check = FALSE
)
```
similarly here :)

```

Therefore, the use of futures and the use of lazy evaluation are orthogonal concepts: you can use future with or without lazy evaluation.

> The future package is about *how* the value is computed (in parallel or sequentially for instance), lazy evaluation is about *when* the value is computed (right as it is defined, or only when it is needed).

Reviewer 2: This is really good and clear!

summary <- mtcars2 %>% 
  group_by(cyl) %>% 
  summarise(mpg = mean(mpg, na.rm = TRUE)) %>% 
  arrange(desc(mpg))

# the object is lazy, the value is not computed yet
summary
nrow(summary)

# we explictly request the data, so now it's there
answer <- collect(summary)
nrow(answer)

```

-   the [dtplyr package](https://dtplyr.tidyverse.org/index.html), which is a data.table back-end for dplyr. The ["lazy" data.table objects](https://dtplyr.tidyverse.org/reference/lazy_dt.html) "captures the intent of dplyr verbs, only actually performing computation when requested" (with `collect()` for instance). The manual also explains that this allows dtplyr to make the code more performant by simplifying the data.table calls.

Slightly tweaked from dtplyr README,

```{r}
# load packages
library(data.table)
library(dtplyr)
library(dplyr, warn.conflicts = FALSE)

# create a “lazy” data table that tracks the operations performed on it.
mtcars2 <- lazy_dt(mtcars)

# create the query
summary <- mtcars2 %>% 
  filter(wt < 5) %>% 
  mutate(l100k = 235.21 / mpg) %>% # liters / 100 km
  group_by(cyl) %>% 
  summarise(l100k = mean(l100k))

# the object is lazy, the value is not computed yet

> summary

Reviewer 2: How does it know what to put into the summary? Something clearly is already there, maybe you can add a sentence on that (possibly even just to point elsewhere for an explanation).

Author: I added a sentence but really I have no idea so the sentence is "here's how it looks like at this stage" :sweat_smile:

nrow(summary)

# we explictly request the data, so now it's there
answer <- as_tibble(summary)
nrow(answer)

```

-   the [duckplyr package](https://duckplyr.tidyverse.org/dev/), which is a drop-in replacement for dplyr, powered by DuckDB for fast operation. Objects created with `duckdb_tibble(.lazy = TRUE)` (`.lazy = TRUE` is not the default) and with `read_parquet_duckdb()` (`lazy=TRUE` is the default) are lazy, the value is only computed if explicitly accessed.

### duckplyr, lazy evaluation and deferred evaluation

> Some doubts to be clarified

Reviewer 2: 👀

## Lazy as in lazy loading of data in packages (`LazyData`)

If your R package exports data, and sets the `LazyData` field in `DESCRIPTION` to `true`, then the exported datasets are lazily loaded: they're available without the use of `data()`, but they're not actually taking up memory until they are accessed.

There's more details on `LazyData` in the [R packages book by Hadley Wickham and Jenny Bryan](https://r-pkgs.org/data.html#sec-data-data) and in [Writing R Extensions](https://cloud.r-project.org/doc/manuals/r-devel/R-exts.html#Data-in-packages).

> Note that internal data is always lazily loaded, and that data that is too big cannot be lazily loaded.

Reviewer 2: can you clarify what "too big" means here?

## Lazy as in frugal file modifications

The `pkgdown::build_site()` function, that creates a documentation website for an R package, features a [`lazy` argument](https://pkgdown.r-lib.org/reference/build_site.html#arg-lazy).
"If `TRUE`, will only rebuild articles and reference pages if the source is newer than the destination."

It is a much simpler concept of laziness: decide right now whether it is needed to rebuild each page.

The potools package, that provides tools for portability and internationalization of R packages, uses ["lazy" for a similar meaning](https://michaelchirico.github.io/potools/reference/po_update.html?q=lazy#ref-usage).

## Lazy as in lazy quantifiers in regular expressions

In regular expressions you can use [quantifiers](https://blog.djnavarro.net/posts/2024-12-16_regex-backreferences/#quantifiers) to indicate how many times a pattern must appear: the pattern can be optional, appear several times, etc.
You can also specify whether the tool should match as many repetitions as possible, or the fewest number of repetitions possible.

Matching the fewest number of repetitions possible is "lazy" (stingy, I made that up).

> Matching as many repetitions as possible is "eager" (or greedy, actual term).

Reviewer 2: Without knowing what the official terms are, this might be slightly confusing? Also, I do see search results with both  lazy and stingy?

Author: and I thought I had made that up  ahahah thanks for checking

```{r}
string <- "aaaaaa"
# greedy! eager!
stringr::str_match(string, "a+")

> # stringy! lazy!

Reviewer 2: ```suggestion
# stingy! lazy!
```

stringr::str_match(string, "a+?")
```

## Conclusion

In the context of lazy evaluation and lazy database operations we can think of lazy as a sort of parcimonious procrastination.
In the case of frugal file modifications in pkgdown and potools, no procrastination, an informed decision is made on the spot on whether a computation is needed.
In the case of lazy quantifiers in regular expressions, lazy means stingy.

> Overall, an user can expect "lazy" means less waste, but it is crucial the documentation of the particular piece of software at hand clarifies the meaning and the potential trade-offs.

Reviewer 2: ```suggestion
Overall, an user can expect "lazy" to mean "less waste", but it is crucial that the documentation of the particular piece of software at hand clarifies the meaning and the potential trade-offs.
```

## Lazy as in frugal file modifications

The `pkgdown::build_site()` function, that creates a documentation website for an R package, features a [`lazy` argument](https://pkgdown.r-lib.org/reference/build_site.html#arg-lazy).
"If `TRUE`, will only rebuild articles and reference pages if the source is newer than the destination."

It is a much simpler concept of laziness: decide right now whether it is needed to rebuild each page.

The potools package, that provides tools for portability and internationalization of R packages, uses ["lazy" for a similar meaning](https://michaelchirico.github.io/potools/reference/po_update.html?q=lazy#ref-usage).

## Lazy as in frugal package testing

The [lazytest package](https://lazytest.cynkra.com/) by Kirill Müller saves you time by only re-running tests that failed during the last run:

-   You run all tests once with `lazytest::lazytest_local()` instead of `devtools::test()`. The lazytest package records which tests failed.
-   The next call to `lazytest::lazytest_local()` only runs the tests that had failed.

This way you can iterate on fixing tests until you get a clean run.

> At which stage it's probably wise to run all tests again to check you didn't break anything else in the meantime.

Reviewer 2: Good reminder!
